In [1]:
%matplotlib inline
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt

# 資料探索

In [2]:
train_data = pd.read_csv('C:/Users/Chun/Desktop/PythonHW/sales_train_v2.csv')
item_data = pd.read_csv('C:/Users/Chun/Desktop/PythonHW/items.csv')

In [3]:
item_data.head()

item_name  item_id  \
0          ! ВО ВЛАСТИ НАВАЖДЕНИЯ (ПЛАСТ.)         D        0   
1  !ABBYY FineReader 12 Professional Edition Full...        1   
2      ***В ЛУЧАХ СЛАВЫ   (UNV)                    D        2   
3    ***ГОЛУБАЯ ВОЛНА  (Univ)                      D        3   
4        ***КОРОБКА (СТЕКЛО)                       D        4   

   item_category_id  
0                40  
1                76  
2                40  
3                40  
4                40

In [4]:
train_data.head()

date  date_block_num  shop_id  item_id  item_price  item_cnt_day
0  02.01.2013               0       59    22154      999.00           1.0
1  03.01.2013               0       25     2552      899.00           1.0
2  05.01.2013               0       25     2552      899.00          -1.0
3  06.01.2013               0       25     2554     1709.05           1.0
4  15.01.2013               0       25     2555     1099.00           1.0

In [5]:
data_sales = pd.merge(train_data, item_data, on = 'item_id')

In [6]:
data_sales.head()

date  date_block_num  shop_id  item_id  item_price  item_cnt_day  \
0  02.01.2013               0       59    22154       999.0           1.0   
1  23.01.2013               0       24    22154       999.0           1.0   
2  20.01.2013               0       27    22154       999.0           1.0   
3  02.01.2013               0       25    22154       999.0           1.0   
4  03.01.2013               0       25    22154       999.0           1.0   

           item_name  item_category_id  
0  ЯВЛЕНИЕ 2012 (BD)                37  
1  ЯВЛЕНИЕ 2012 (BD)                37  
2  ЯВЛЕНИЕ 2012 (BD)                37  
3  ЯВЛЕНИЕ 2012 (BD)                37  
4  ЯВЛЕНИЕ 2012 (BD)                37

In [56]:
data_sales['item_cnt_day'].mode()

0    1.0
dtype: float64

In [52]:
data_sales[data_sales['item_cnt_day']==-22]

date  date_block_num  shop_id  item_id  item_price  item_cnt_day  \
906649 2013-02-11              10       12     8023        15.0         -22.0   

                                         item_name  item_category_id  day  \
906649  ZL004 Лазерные наклейки Смайлики маленькие                66   11   

        Month  Year  
906649      2  2013

In [7]:
data_sales.shape

(2935849, 8)

# 資料格式整理

In [8]:
data_sales["date"] = pd.to_datetime(data_sales["date"])

In [9]:
#每一個月(0~33)每一個 shop_id 的每一個 item_id 的銷售商品總數
#data_sales.groupby(['date_block_num','shop_id', 'item_id'])['item_cnt_day'].sum()

將date的格式轉成時間格式

In [10]:
data_sales.sort_values(by = ['shop_id','item_id', 'date']).tail()

date  date_block_num  shop_id  item_id  item_price  \
2781093 2015-11-02              25       59    22164       749.0   
1850368 2013-03-12              11       59    22167       299.0   
1850219 2013-10-25               9       59    22167       299.0   
1850367 2013-12-14              11       59    22167       299.0   
1850642 2014-06-14              17       59    22167       299.0   

         item_cnt_day                                          item_name  \
2781093           1.0                                        ЯРОСТЬ (BD)   
1850368           1.0  Язык запросов 1С:Предприятия 8 (+CD). Хрустале...   
1850219           1.0  Язык запросов 1С:Предприятия 8 (+CD). Хрустале...   
1850367           1.0  Язык запросов 1С:Предприятия 8 (+CD). Хрустале...   
1850642           1.0  Язык запросов 1С:Предприятия 8 (+CD). Хрустале...   

         item_category_id  
2781093                37  
1850368                49  
1850219                49  
1850367                49  
1850642                49

In [11]:
data_sales['date'].max()

Timestamp('2015-12-10 00:00:00')

將資料中 shop_id 和 item_id 中所有可能的組合抓出來

In [12]:
shop_item = data_sales[['shop_id','item_id']]

In [13]:
shop_item.head()

shop_id  item_id
0       59    22154
1       24    22154
2       27    22154
3       25    22154
4       25    22154

In [14]:
shop_item.drop_duplicates(inplace = True)

C:\Users\Chun\Anaconda3\lib\site-packages\ipykernel_launcher.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy
  """Entry point for launching an IPython kernel.


In [15]:
shop_item.head()

shop_id  item_id
0       59    22154
1       24    22154
2       27    22154
3       25    22154
8        6    22154

In [16]:
shop_item

shop_id  item_id
0             59    22154
1             24    22154
2             27    22154
3             25    22154
8              6    22154
9             15    22154
10             7    22154
11            31    22154
12            54    22154
13            42    22154
14            38    22154
15            37    22154
16            46    22154
17            44    22154
18            51    22154
20            16    22154
21             1    22154
22             0    22154
24            28    22154
25            57    22154
26            23    22154
28            58    22154
31            56    22154
40            35    22154
46            47    22154
48            21    22154
59            25     2552
61             6     2552
62            30     2552
63             4     2552
...          ...      ...
2935816       25    13245
2935817       31     5953
2935818       25     5953
2935820       31     5959
2935821       31    10935
2935822       31    10062
2935823       31      238
2935824       25      238
2935825       31      986
2935826       31      461
2935827       21    15889
2935828       21    16093
2935829       21    15417
2935830       21    21073
2935831       25      139
2935832       25      315
2935833       25      445
2935834       25      930
2935835       22     7151
2935836       22    15414
2935837       25    13476
2935840       25    13280
2935841       25    16105
2935842       25    16094
2935843       25    14709
2935844       25     8428
2935845       25     7903
2935846       25     7610
2935847       25     7635
2935848       25     7640

[424124 rows x 2 columns]

先複製一份原始資料，將日期格式分成日、月、年合併到data_sales_copy裡

In [17]:
data_sales_copy = data_sales

In [18]:
data_sales_copy['day'] = data_sales_copy['date'].dt.day

In [19]:
data_sales_copy['Month'] = data_sales_copy['date'].dt.month

In [20]:
data_sales_copy['Year'] = data_sales_copy['date'].dt.year

input檔案格式解釋：

第一列：
shop_id = 59, item_id = 22154 在 2013 年 1 月，每一天的銷售量[0,0,0,....,0] (長度為31) (1月沒有賣出)


第二列：
shop_id = 59, item_id = 22154 在 2013 年 2 月，每一天的銷售量[1,0,0,....,0] (長度為31) (只有2/1有賣出)
...
...


以此類推，每一間商店每一項商品都會有34列的資料(表示2013年1月到2015年10月)

In [21]:
#2013年shop_item[0]的組合

everydaySalesInMonth = np.zeros(31)
day = data_sales_copy.loc[(data_sales_copy['Year'] == 2013) & (data_sales_copy['Month'] == 1) & (data_sales_copy['shop_id'] == shop_item.loc[0][0]) & (data_sales_copy['item_id'] == shop_item.loc[0][1]),'day']
everydaySalesInMonth[day - 1] = data_sales_copy.loc[(data_sales_copy['Year'] == 2013) & (data_sales_copy['Month'] == 1) & (data_sales_copy['shop_id'] == shop_item.loc[0][0]) & (data_sales_copy['item_id'] == shop_item.loc[0][1]),'item_cnt_day']
X_train = everydaySalesInMonth.reshape(1,31)
X_train = pd.DataFrame(X_train)



for i in range(2,13):
    everydaySalesInMonth = np.zeros(31)
    day = data_sales_copy.loc[(data_sales_copy['Year'] == 2013) & (data_sales_copy['Month'] == i) & (data_sales_copy['shop_id'] == shop_item.loc[0][0]) & (data_sales_copy['item_id'] == shop_item.loc[0][1]),'day']
    everydaySalesInMonth[day - 1] = data_sales_copy.loc[(data_sales_copy['Year'] == 2013) & (data_sales_copy['Month'] == i) & (data_sales_copy['shop_id'] == shop_item.loc[0][0]) & (data_sales_copy['item_id'] == shop_item.loc[0][1]),'item_cnt_day']
    everydaySalesInMonth = everydaySalesInMonth.reshape(1,31)
    everydaySalesInMonth = pd.DataFrame(everydaySalesInMonth)
    X_train = pd.concat([X_train, pd.DataFrame(everydaySalesInMonth)],axis=0)

In [22]:
X_train

0    1    2    3    4    5    6    7    8    9  ...    21   22   23   24  \
0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0 ...   0.0  0.0  0.0  0.0   
0  1.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0 ...   0.0  0.0  0.0  0.0   
0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0 ...   0.0  0.0  0.0  0.0   
0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0 ...   0.0  0.0  0.0  0.0   
0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0 ...   0.0  0.0  0.0  0.0   
0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0 ...   0.0  0.0  0.0  0.0   
0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0 ...   0.0  0.0  0.0  0.0   
0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0 ...   0.0  0.0  0.0  0.0   
0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0 ...   0.0  0.0  0.0  0.0   
0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0 ...   0.0  0.0  0.0  0.0   
0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0 ...   0.0  0.0  0.0  0.0   
0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0 ...   0.0  0.0  0.0  0.0   

    25   26   27   28   29   30  
0  0.0  0.0  0.0  0.0  0.0  0.0  
0  0.0  0.0  0.0  0.0  0.0  0.0  
0  0.0  0.0  0.0  0.0  0.0  0.0  
0  0.0  0.0  0.0  0.0  0.0  0.0  
0  0.0  0.0  0.0  0.0  0.0  0.0  
0  0.0  0.0  0.0  0.0  0.0  0.0  
0  0.0  0.0  0.0  0.0  0.0  0.0  
0  0.0  0.0  0.0  0.0  0.0  0.0  
0  0.0  0.0  0.0  0.0  0.0  0.0  
0  0.0  0.0  0.0  0.0  0.0  0.0  
0  0.0  0.0  0.0  0.0  0.0  0.0  
0  0.0  0.0  0.0  0.0  0.0  0.0  

[12 rows x 31 columns]

In [57]:
x_train = X_train.values
#y_train = Y_train.values
y_train = [1,2,3,4,5,5,4,3,2,1,0,0]

In [58]:
from keras.preprocessing import sequence
from keras.models import Sequential
from keras.layers import Dense,Embedding, LSTM, Dropout

In [59]:
x_element_MAX_Dimention = 2169+22
N=10 #2169+22要壓到幾維?
K=10 #LSTM的個數

In [60]:
model = Sequential()

In [61]:
model.add(Embedding(x_element_MAX_Dimention,N)) 
model.add(LSTM(K))
model.add(Dense(1, activation='sigmoid'))
model.compile(loss='mse', optimizer='adam', metrics=['accuracy'])
model.summary()

_________________________________________________________________
Layer (type)                 Output Shape              Param #   
embedding_6 (Embedding)      (None, None, 10)          21910     
_________________________________________________________________
lstm_7 (LSTM)                (None, 10)                840       
_________________________________________________________________
dense_7 (Dense)              (None, 1)                 11        
Total params: 22,761
Trainable params: 22,761
Non-trainable params: 0
_________________________________________________________________


In [62]:
model.fit(x_train, y_train, batch_size=100, epochs=2)

Epoch 1/2
12/12 [==============================] - 1s 73ms/step - loss: 6.9238 - acc: 0.1667
Epoch 2/2
12/12 [==============================] - 0s 914us/step - loss: 6.9167 - acc: 0.1667
